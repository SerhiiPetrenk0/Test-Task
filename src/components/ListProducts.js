import React, { useEffect, useState } from 'react';
import { Container } from 'react-bootstrap';
import { getProductsAPI } from '../api'
import * as Styled from '../styled/ListProducts';

export const ListProducts = () => {
    const [List, setList] = useState([]);

    const renderItems = List.map(item => (
          <Styled.Col key={item.asin} lg="3" md="4" sm="6">
          <Styled.Link key={item.asin} to={`/Product/${item.asin}`}>
            <Styled.Card>
              <Styled.CardImg variant="top" src={item.img}/>
              <Styled.CardBody>
                <Styled.CardTitle>{item.brand}</Styled.CardTitle>
                <Styled.CardSubtitle>{item.price}$</Styled.CardSubtitle>
                <Styled.CardText>
                  {item.name}
                </Styled.CardText>
              </Styled.CardBody>
            </Styled.Card>
          </Styled.Link>
          </Styled.Col>
    ));

    useEffect(() => {
      getProductsAPI(setList)
    }, []);

    return (
        <Container>
          <Styled.Row>
            {renderItems}
          </Styled.Row>
        </Container>
    );
}
