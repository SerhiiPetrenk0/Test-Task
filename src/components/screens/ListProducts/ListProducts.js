import React, { useEffect, useState } from 'react';
import { Container } from 'react-bootstrap';
import { getProductsAPI } from '../../../api'
import {
  StyledCol,
  StyledLink, 
  StyledCard,
  StyledCardImg,
  StyledCardBody,
  StyledCardTitle,
  StyledCardSubtitle,
  StyledCardText,
  StyledRow
} from '../../../styled/screens/ListProducts/ListProducts';

export const ListProducts = () => {

    const [List, setList] = useState([]);

    const renderItems = List.map(item => (
          <StyledCol key={item.asin} lg="3" md="4" sm="6">
          <StyledLink key={item.asin} to={`/Product/${item.asin}`}>
            <StyledCard>
              <StyledCardImg variant="top" src={item.img}/>
              <StyledCardBody>
                <StyledCardTitle>{item.brand}</StyledCardTitle>
                <StyledCardSubtitle>{item.price}$</StyledCardSubtitle>
                <StyledCardText>
                  {item.name}
                </StyledCardText>
              </StyledCardBody>
            </StyledCard>
          </StyledLink>
          </StyledCol>
    ));

    useEffect(() => {
      getProductsAPI(setList)
    }, []);

    return (
        <Container>
          <StyledRow>
            {renderItems}
          </StyledRow>
        </Container>
    );
}