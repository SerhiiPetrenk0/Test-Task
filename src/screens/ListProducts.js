import React, { useEffect, useState } from 'react';
import { ListGroup } from 'react-bootstrap';

export const ListProducts = () => {

    const [List, setList] = useState([]);
    const loadItems = () => {
      fetch('http://demo8021751.mockable.io/products')
        .then((response) => {
          return response.json();
        })
        .then((data) => {
          setList(data.products);
        });
    };
    const renderItems = List.map((item, index) => (
      <ListGroup.Item key={index}> {item.brand} </ListGroup.Item>
    ));
  
    useEffect(() => {
      loadItems();
    }, []);
    return (
        <ListGroup>{renderItems}</ListGroup>
    );
}