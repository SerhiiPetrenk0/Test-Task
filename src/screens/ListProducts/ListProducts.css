.item {
    background-color: rgb(255, 255, 255);
    transition: .3s;
    cursor: pointer;
}
.link {
    text-decoration: none;
}
.item:hover {
    background-color: rgba(223, 223, 223, 0.356);
}
.item-img {
    height: 150px;
}