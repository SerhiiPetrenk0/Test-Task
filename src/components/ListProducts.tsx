import React, { useEffect, useState } from 'react';
import { Container } from 'react-bootstrap';
import * as Styled from '../styled/ListProducts';
import { loadProduct } from '../redux/ducks/productsDuck';
import { useSelector, useDispatch } from 'react-redux';
import { TypeProduct } from '../interface';

export const ListProducts:React.FC = () => {
    const [List, setList] = useState<TypeProduct[]>([]);
    const dispatch = useDispatch();
    type TStore = {
      product: {
        products:TypeProduct[]
      }
    };
    const store:TypeProduct[] = useSelector((store: TStore) => store.product.products);

    const renderItems:JSX.Element[] = List.map((item: TypeProduct) => (
          <Styled.Col key={item.asin} lg="3" md="4" sm="6">
            <Styled.Link key={item.asin} to={`/Product/${item.asin}`}>
              <Styled.Card>
                <Styled.CardImg variant="top" src={item.img}/>
                <Styled.CardBody>
                  <Styled.CardTitle>{item.brand}</Styled.CardTitle>
                  <Styled.CardSubtitle>{item.price}$</Styled.CardSubtitle>
                  <Styled.CardText>
                    {item.name}
                  </Styled.CardText>
                </Styled.CardBody>
              </Styled.Card>
            </Styled.Link>
          </Styled.Col>
    ));

    const checkStore: () => void = () => store.length === 0 ? dispatch(loadProduct()) : setList(store);

    useEffect(() => {
      checkStore();
    });

    return (
      <>
        {!!!store ? (
            <Styled.Spinner animation="border" /> 
          ) : (
            <Container>
              <Styled.Row>
                {renderItems}
              </Styled.Row>
            </Container>
        )}
      </>
    );
};
