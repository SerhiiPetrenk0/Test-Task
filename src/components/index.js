//Screens
import { ListProducts } from './ListProducts';
import { ItemProducts } from './ItemProducts';
//Components
import { Comment } from './components/Comment';
import { PostForm } from './components/PostForm';
import { StarBar } from './components/StarBar';

export { 
    ListProducts,
    ItemProducts, 
    Comment, 
    PostForm, 
    StarBar 
};
