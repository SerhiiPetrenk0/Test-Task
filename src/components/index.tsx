//Screens
import { ListProducts } from './ListProducts';
import { ItemProducts } from './ItemProducts';
//Components
import { Comment } from './other/Comment';
import { PostForm } from './other/PostForm';
import { StarBar } from './other/StarBar';

export { 
    ListProducts,
    ItemProducts, 
    Comment, 
    PostForm, 
    StarBar 
};
