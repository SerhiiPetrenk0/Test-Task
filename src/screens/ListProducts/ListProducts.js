import React, { useEffect, useState } from 'react';
import { ListGroup, Container, Row, Col, Image } from 'react-bootstrap';
import "./ListProducts.css"

export const ListProducts = () => {

    const [List, setList] = useState([]);
    const loadItems = () => {
      fetch('http://demo8021751.mockable.io/products')
        .then((response) => {
          return response.json();
        })
        .then((data) => {
          setList(data.products);
        });
    };
    const renderItems = List.map((item, index) => (
      <ListGroup.Item className='item m-2 border border-1 rounded-3' key={index}>
        <Row className='p-2 d-flex justify-content-center'>
          <Col xs={6} md={4} className="item-img d-block position-relative" >
            <Image src={item.img} className="mh-100 mw-100 position-absolute top-50 start-50 translate-middle" alt="***" />
          </Col>
          <Col xs={12} md={8}>
            <Row><h4 className="display-6 mt-1 text-center text-md-start">{item.brand}</h4></Row>
            <Row><h4 className='text-center mt-4 text-md-start'>{item.price}$</h4></Row>
          </Col>
        </Row>
      </ListGroup.Item>
    ));
  
    useEffect(() => {
      loadItems();
    }, []);
    return (
        <Container className="justify-content-md-center">
          <ListGroup>{renderItems}</ListGroup>
        </Container>
    );
}