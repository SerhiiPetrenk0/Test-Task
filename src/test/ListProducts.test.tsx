import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import * as reactRedux from 'react-redux'
import 'jsdom-global/register';

import { ListProducts } from '../components/ListProducts';

configure({ adapter: new Adapter() });
const initialStore = [
    {
    in_potential_products: false,
    asin: "B0721KZHWT",
    price: 15.79,
    weight: 0.61949822,
    updated_at: 1537974537000.0,
    name: "A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar\/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included",
    shipping_weight: null,
    domain: "uk",
    votes_count: 51,
    daily_cashflow: 521.07,
    currency: "GBP",
    img: "https://m.media-amazon.com/images/I/71FG+7OuuKL._AC_SX466_.jpg",
    daily_sales: 33,
    stars: 4.6,
    bsr_value: 2360,
    brand_class: null,
    size: [
        0.0393701,
        11.417329,
        16.535442
    ],
    link: "https:\/\/www.amazon.co.uk\/Magnetic-Weekly-Planner-Whiteboard-Fridge\/dp\/B0721KZHWT\/ref=sr_1_17?ie=UTF8&qid=1537981729&sr=8-17&keywords=fridge+organisers",
    bsr_category: "Home & Kitchen",
    brand: "Plan Smart"
},
{
    in_potential_products: false,
    asin: "B0721KFGWT",
    price: 15.79,
    weight: 0.61949822,
    updated_at: 1537974537000.0,
    name: "A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar\/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included",
    shipping_weight: null,
    domain: "uk",
    votes_count: 51,
    daily_cashflow: 521.07,
    currency: "GBP",
    img: "https://m.media-amazon.com/images/I/71FG+7OuuKL._AC_SX466_.jpg",
    daily_sales: 33,
    stars: 4.6,
    bsr_value: 2360,
    brand_class: null,
    size: [
        0.0393701,
        11.417329,
        16.535442
    ],
    link: "https:\/\/www.amazon.co.uk\/Magnetic-Weekly-Planner-Whiteboard-Fridge\/dp\/B0721KZHWT\/ref=sr_1_17?ie=UTF8&qid=1537981729&sr=8-17&keywords=fridge+organisers",
    bsr_category: "Home & Kitchen",
    brand: "Plan Smart"
},
];

describe('ListProducts component', () => {
  const myInitialState = initialStore;
  React.useState = jest.fn().mockReturnValue([myInitialState, {}]);
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');
  const dummyDispatch = jest.fn();

  beforeEach(() => {
      useSelectorMock.mockClear();
      useDispatchMock.mockClear();
    });

  useSelectorMock.mockReturnValue(initialStore);
  useDispatchMock.mockReturnValue(dummyDispatch);

  it('Snapshot ListProducts component', () => {
    const wrapper = shallow(<ListProducts />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
