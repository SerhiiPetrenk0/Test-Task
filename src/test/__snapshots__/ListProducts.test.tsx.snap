// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListProducts component Snapshot ListProducts component 1`] = `
<Fragment>
  <Container
    fluid={false}
  >
    <Styled(Row)>
      <Styled(Col)
        key="B0721KZHWT"
        lg="3"
        md="4"
        sm="6"
      >
        <Styled(Link)
          key="B0721KZHWT"
          to="/Product/B0721KZHWT"
        >
          <Styled(Card)>
            <Styled(CardImg)
              src="https://m.media-amazon.com/images/I/71FG+7OuuKL._AC_SX466_.jpg"
              variant="top"
            />
            <Styled(CardBody)>
              <Styled(CardTitle)>
                Plan Smart
              </Styled(CardTitle)>
              <Styled(CardSubtitle)>
                15.79
                $
              </Styled(CardSubtitle)>
              <Styled(CardText)>
                A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included
              </Styled(CardText)>
            </Styled(CardBody)>
          </Styled(Card)>
        </Styled(Link)>
      </Styled(Col)>
      <Styled(Col)
        key="B0721KFGWT"
        lg="3"
        md="4"
        sm="6"
      >
        <Styled(Link)
          key="B0721KFGWT"
          to="/Product/B0721KFGWT"
        >
          <Styled(Card)>
            <Styled(CardImg)
              src="https://m.media-amazon.com/images/I/71FG+7OuuKL._AC_SX466_.jpg"
              variant="top"
            />
            <Styled(CardBody)>
              <Styled(CardTitle)>
                Plan Smart
              </Styled(CardTitle)>
              <Styled(CardSubtitle)>
                15.79
                $
              </Styled(CardSubtitle)>
              <Styled(CardText)>
                A3 Magnetic Dry Wipe Weekly Planner Whiteboard for Fridge - Handy Week Calendar/Organiser by Plan Smart - Bonus: 3 Quality Dry Erase Markers with Eraser Included
              </Styled(CardText)>
            </Styled(CardBody)>
          </Styled(Card)>
        </Styled(Link)>
      </Styled(Col)>
    </Styled(Row)>
  </Container>
</Fragment>
`;
